### PR TITLE
[core][runtime_env] Retry on read error in rt env agent client response read

### LINKED ISSUE #45353: [core] Raylet should wait for RuntimeEnvAgent to start before receiving tasks.
### What happened + What you expected to happen

In a node start up:

1. raylet starts
2. raylet spawns RuntimeEnvAgent
3. raylet spawns workers
4. raylet start receiving tasks, assigning to workers

Here we have an race condition: (2) should happen-before (4) but we don't have any code to do the waiting. If the tasks are received and started before the runtime env agent is ready to receive requests, we can have runtime env set up failures.

Note the waiting is not "process started" but needs to go through a HTTP probe. This means we need:

1. an extra endpoint in RE Agent `GET /ping` which returns whatever.
2. raylet polls every 1s until got a 200 OK.

### Versions / Dependencies

master

### Reproduction script

N/A

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by rynewang**
#45353 reports issues on runtime env agent start up process + task pressure. Some times we get `on_read` errors. This PR fixes by retrying on those requests. Specifically, all HTTP errors before we get a well-formed HTTP response are retried transparently with a ray log.

Also I took this chance to do some quality-of-life improvements to the runtime env agent:

- Moved forward the log initialization, so we have "process started" logs and "ready to serve traffic" logs separately.
- Allow the http client to set method.
- Explicitly set HTTP/1.1.
- Uses Beast's way to set content-length.
- Sets never expire on each step.
- Returns NotFound on resolving and connecting, returns Disconnected on on_write and on_read, returns IOError on user error (HTTP non OK).
- Retry on NotFound and Disconnected.

Added a unit test for this situation.

Fixes https://github.com/ray-project/ray/issues/45353.

### CODE REVIEW

**Comment by jjyao** [comment]
A high level question: is it guaranteed to be safe to retry upon `on_read` errors?

**Review by hongchaodeng — APPROVED** [review]
LGTM after nit

**Comment by hongchaodeng** [comment]
I have reproduced the issue and verified that it is safe to retry when `on_read` failed.

Here's the log from runtime env agent:
```
2024-06-03 20:33:53,759 INFO runtime_env_agent.py:380 -- Creating runtime env: {"_ray_commit": "{{RAY_COMMIT_SHA}}", "pip": {"packages": ["emoji"], "pip_check": false}} with timeout 600 seconds.
2024-06-03 20:33:57,751 INFO runtime_env_agent.py:428 -- Successfully created runtime env: {"_ray_commit": "{{RAY_COMMIT_SHA}}", "pip": {"packages": ["emoji"], "pip_check": false}}, the context: {"command_prefix": ["source", "/tmp/ray/session_2024-06-03_20-25-13_549115_257179/runtime_resources/pip/69cd483986f3d2762be1e312f9cd49dac02e9f97/virtualenv/bin/activate", "1>&2", "&&"], "env_vars": {}, "py_executable": "/tmp/ray/session_2024-06-03_20-25-13_549115_257179/runtime_resources/pip/69cd483986f3d2762be1e312f9cd49dac02e9f97/virtualenv/bin/python", "resources_dir": null, "override_worker_entrypoint": null, "java_jars": []}
2024-06-03 20:33:57,751 INFO runtime_env_agent.py:464 -- Runtime env already created successfully. Env: {"_ray_commit": "{{RAY_COMMIT_SHA}}", "pip": {"packages": ["emoji"], "pip_check": false}}, context: {"command_prefix": ["source", "/tmp/ray/session_2024-06-03_20-25-13_549115_257179/runtime_resources/pip/69cd483986f3d2762be1e312f9cd49dac02e9f97/virtualenv/bin/activate", "1>&2", "&&"], "env_vars": {}, "py_executable": "/tmp/ray/session_2024-06-03_20-25-13_549115_257179/runtime_resources/pip/69cd483986f3d2762be1e312f9cd49dac02e9f97/virtualenv/bin/python", "resources_dir": null, "override_worker_entrypoint": null, "java_jars": []}
2024-06-03 20:33:58,753 INFO runtime_env_agent.py:464 -- Runtime env already created successfully. Env: {"_ray_commit": "{{RAY_COMMIT_SHA}}", "pip": {"packages": ["emoji"], "pip_check": false}}, context: {"command_prefix": ["source", "/tmp/ray/session_2024-06-03_20-25-13_549115_257179/runtime_resources/pip/69cd483986f3d2762be1e312f9cd49dac02e9f97/virtualenv/bin/activate", "1>&2", "&&"], "env_vars": {}, "py_executable": "/tmp/ray/session_2024-06-03_20-25-13_549115_257179/runtime_resources/pip/69cd483986f3d2762be1e312f9cd49dac02e9f97/virtualenv/bin/python", "resources_dir": null, "override_worker_entrypoint": null, "java_jars": []}
```

It will reply STATUS_OK with correct context.

File: python/ray/tests/test_runtime_env.py [hunk 1]
```diff
@@ -388,6 +388,62 @@ def f():
...
+    indirect=True,
+)
+def test_slow_runtime_env_agent_startup_on_task_pressure(
+    shutdown_only, set_runtime_env_plugins
+):
+    """
+    Starts nodes with runtime env agent and a slow plugin. Then when the runtime env
+    agent is still starting up, we submit a lot of tasks to the cluster. The tasks
+    should wait for the runtime env agent to start up and then run.
+    https://github.com/ray-project/ray/issues/45353
```
Comment (jjyao) [inline python/ray/tests/test_runtime_env.py:426]: This is not accurate? since we won't wait but just retry

**Review by jjyao — APPROVED** [review]

File: python/ray/tests/test_runtime_env.py (hunk 1 above)
Comment (rynewang) [reply python/ray/tests/test_runtime_env.py:426]: we wait every 1s before retry

Comment (rynewang) [reply python/ray/tests/test_runtime_env.py:426]: Actually waits for agent_manager_retry_interval_ms_ = 100ms, until total = agent_register_timeout_ms = 10000ms = 10s and hard error out.

**Comment by rynewang** [comment]
Test timeout fixed - it turns out one previous test set some timeout-ing os env var and did not set back. Changed all such env vars to monkeypatch.

**Pull request merged**